### PR TITLE
[#IOCIT-341] Implement GetAssertionHandler

### PR DESCRIPTION
**Pull request by gquadrati**
#### List of Changes

Add `GetAssertionHandler` implementation

#### Motivation and Context

Enable LolliPoP Consumer to retrieve an assertion

#### How Has This Been Tested?

Unit tests, Integration tests

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: utils/readers.ts [hunk 1]
```diff
@@ -1,19 +1,37 @@
...
   RetrievedLolliPopPubKeys
 >;
 
+export type AssertionReader = RTE.ReaderTaskEither<
+  AssertionFileName,
+  DomainError,
+  NonEmptyString
+>;
+
 // IMPLEMENTATIONS
```
Comment (michaeldisaro) [inline utils/readers.ts:35]: imho this comment is useless.

File: utils/readers.ts [hunk 2]
```diff
@@ -28,3 +46,24 @@ export const getPopDocumentReader = (
     })),
     TE.chainW(TE.fromOption(() => ({ kind: ErrorKind.NotFound as const })))
   );
+
+// IMPLEMENTATIONS
```
Comment (michaeldisaro) [inline utils/readers.ts:50]: ditto

**Review by michaeldisaro — APPROVED** [review]

File: __integrations__/__tests__/getAssertion.test.ts [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+): Promise<AssertionRef> => {
+  const thumbprint = await jose.calculateJwkThumbprint(jwk, algo);
+  return `${algo}-${thumbprint}` as AssertionRef;
+};
+
+// -------------------------
+// Tests
+// -------------------------
+
+describe("getAssertion |> Validation Failures", () => {
```
Comment (arcogabbo) [inline __integrations__/__tests__/getAssertion.test.ts:144]: can we add a describe for the error `410` when a valid payload is sent but the `popDocument` or `assertion` are not found?

File: __integrations__/environments/env.base [hunk 4]
```diff
@@ -26,3 +26,4 @@ SHOW_LOGS=false
 # Test Data
 # ----------
 
+BEARER_AUTH_HEADER=x-pagopa-lollipop-auth
```
Comment (arcogabbo) [inline __integrations__/environments/env.base:29]:
```suggestion
BEARER_AUTH_HEADER=x-pagopa-lollipop-auth-jwt
```
shouldn't this be `x-pagopa-lollipop-auth-jwt`?

File: __integrations__/__tests__/getAssertion.test.ts [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+const LOLLIPOP_ASSERTION_STORAGE_CONTAINER_NAME = "assertions";
+
+// ----------------
+// Setup dbs
+// ----------------
+
+const blobService = createBlobService(QueueStorageConnection);
+
+// @ts-ignore
```
Comment (arcogabbo) [inline __integrations__/__tests__/getAssertion.test.ts:81]: not needed

File: utils/readers.ts (hunk 1 above)
Comment (gquadrati) [reply utils/readers.ts]: Done in [80a493d](https://github.com/pagopa/io-functions-lollipop/pull/26/commits/80a493db6bd5b57663e3f88ef3461acf83972ec1)

File: utils/readers.ts (hunk 2 above)
Comment (gquadrati) [reply utils/readers.ts]: Done in [80a493d](https://github.com/pagopa/io-functions-lollipop/pull/26/commits/80a493db6bd5b57663e3f88ef3461acf83972ec1)

File: GetAssertion/handler.ts [hunk 6]
```diff
@@ -38,31 +53,74 @@ import { LCUserInfo } from "../generated/definitions/external/LCUserInfo";
...
+    TE.fromPredicate(
+      ar => ar === jwt.assertionRef,
+      () =>
+        logAndReturnResponse(
+          context,
+          ResponseErrorForbiddenNotAuthorized,
+          `jwt assertion_ref does not match the one in path`
+        )
+    ),
+    TE.chainW(
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply GetAssertion/handler.ts]: Done in [80a493d](https://github.com/pagopa/io-functions-lollipop/pull/26/commits/80a493db6bd5b57663e3f88ef3461acf83972ec1)

File: GetAssertion/handler.ts [hunk 7]
```diff
@@ -71,15 +129,16 @@ export const GetAssertionHandler = (
  */
 // eslint-disable-next-line max-params, prefer-arrow/prefer-arrow-functions
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply GetAssertion/handler.ts]: Done in [80a493d](https://github.com/pagopa/io-functions-lollipop/pull/26/commits/80a493db6bd5b57663e3f88ef3461acf83972ec1)

File: __integrations__/__tests__/getAssertion.test.ts [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+    expect(response.status).toEqual(403);
+    const body = await response.json();
+    expect(body).toMatchObject({
+      status: 403,
+      title: "You are not allowed here",
+      detail: `You do not have enough permission to complete the operation you requested`
+    });
+  });
+
+  it("should fail when the jwt expired", async () => {
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply __integrations__/__tests__/getAssertion.test.ts]: Done in [80a493d](https://github.com/pagopa/io-functions-lollipop/pull/26/commits/80a493db6bd5b57663e3f88ef3461acf83972ec1)

File: __integrations__/environments/env.base (hunk 4 above)
Comment (gquadrati) [reply __integrations__/environments/env.base:29]: We defined it `x-pagopa-lollipop-auth` in [openapi param](https://github.com/pagopa/io-functions-lollipop/blob/558f8899b905bbbaa828b566726de81a156d8513/openapi/external.yaml#L65) and [config](https://github.com/pagopa/io-functions-lollipop/blob/558f8899b905bbbaa828b566726de81a156d8513/utils/config.ts#L74)

File: __integrations__/__tests__/getAssertion.test.ts [hunk 9]
```diff
@@ -247,6 +249,61 @@ describe("getAssertion |> Validation Failures", () => {
...
+    const body = await response.json();
+    expect(body).toMatchObject({
+      detail: "Resource gone"
+    });
+  });
+
+  it("should fail when the assertion cannot be found in Blob Storage", async () => {
+    const lcParams = await setupTestAndGenerateLcParams();
+
+    // Recreate the DB to clean-up data
```
Comment (arcogabbo) [inline __integrations__/__tests__/getAssertion.test.ts:281]:
```suggestion
    // Delete Blob to let retrieve fail later in the flow
```

File: __integrations__/__tests__/getAssertion.test.ts (hunk 3 above)
Comment (gquadrati) [reply __integrations__/__tests__/getAssertion.test.ts:145]: Done in [437edfd](https://github.com/pagopa/io-functions-lollipop/pull/26/commits/437edfd1a5af541b2131040944239645491b3a3a)

**Review by arcogabbo — APPROVED** [review]

File: __integrations__/utils/client.ts [hunk 10]
```diff
@@ -49,3 +49,20 @@ export const fetchGenerateLcParams = (
       body: JSON.stringify(body)
     }
   );
+
+export const GET_ASSERTION_PATH = "api/v1/assertions";
+export const fetchGetAssertion = (
+  assertionRef: string,
+  jwtHeaderName: string,
+  jwt: string,
```
Comment (BurnedMarshal) [inline __integrations__/utils/client.ts:57]: to handle missing jwt this is optional
```suggestion
  jwt?: string,
```

**Pull request merged**